### PR TITLE
[stm] Remove Opacity Information from Theorem and Definition start.

### PR DESCRIPTION
**Pull request by ejgallego**
This information was introduced in db65876 and IMHO was wrongly used all the way around.

It seems that some features of the proof layer and the document/STM ought to be better integrated; in particular as discussed in #682 we may want to revert eef907e and in fact go back to a design where the proof terms a proof is allowed to produce correspond to the information stored in `VtStartProof`.

I agree with the spirit of db65876 in the sense that the current static analysis is tied to document
traversal and this is inflexible in some scenarios; however we may want to be more regular at the vernac level too.

This PR is more a RFC than a request for immediate merge, and it depends on #1089

### CODE REVIEW

File: stm/vernac_classifier.ml [hunk 1]
```diff
@@ -66,10 +60,7 @@ let rec classify_vernac e =
     (* Nested vernac exprs *)
     | VernacProgram e -> classify_vernac e
     | VernacLocal (_,e) -> classify_vernac e
-    | VernacPolymorphic (b, e) -> 
-      if b || Flags.is_universe_polymorphism () (* Ok or not? *) then
-	make_polymorphic (classify_vernac e)
-      else classify_vernac e
+    | VernacPolymorphic (b, e) -> classify_vernac e
```
Comment (gares) [inline stm/vernac_classifier.ml]: how is this related to this PR?

File: stm/stm.ml [hunk 2]
```diff
@@ -2731,12 +2729,7 @@ let process_transaction ?(newtip=Stateid.fresh ())
             if not in_proof && Proof_global.there_are_pending_proofs () then
             begin
               let bname = VCS.mk_branch_name x in
-              let rec opacity_of_produced_term = function
-                (* This AST is ambiguous, hence we check it dynamically *)
-                | VernacInstance (false, _,_ , None, _) -> GuaranteesOpacity
-                | VernacLocal (_,e) -> opacity_of_produced_term e
-                | _ -> Doesn'tGuaranteeOpacity in
-              VCS.commit id (Fork (x,bname,opacity_of_produced_term x.expr,[]));
+              VCS.commit id (Fork (x,bname,[]));
```
Comment (gares) [inline stm/stm.ml]: I'm puzzled @maximedenes `under_control` is not used here. For the rest, this is OK

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
I approve the RFC, the poly thing is suspicious.. is it intentional?

File: stm/vernac_classifier.ml (hunk 1 above)
Comment (ejgallego) [reply stm/vernac_classifier.ml]: Look above, `make_polymorphic` is gone. What was that supposed to mean?

**Comment by aspiwack** [comment]
I really have no idea what's the problem here.

But things that seem to make the `Derive` plugin impossible make me suspicious: `Derive` has a perfectly clear semantics. I'd rather we don't throw it away in order for Coq to fit a narrow model of a proof assistant, but that instead we use such a plugin as a constraint on design to help do The Right Thing on the backend side.

Specifically, about this and about #682, I have no idea what the problem is (and I'm guessing I'm not alone). Why this PR can be seen as an good solution. And certainly not what would make it a good trade-off to throw a feature away. Anything that takes features off of Coq should be very careful in explaining the trade-offs.

**Comment by ejgallego** [comment]
This doesn't make `Derive` impossible, it just removes an ad hoc mechanism introduced for it. `Derive` should just be marked as "never process in parallel", and that'd be good.

Regarding the removal of `Derive`, there is quite a bit of overlap between `proof` and `stm` to clean up. In that sense I'd rather advocate a simple semantics for now: "proof starting command exactly produce one proof term". More exotic solutions would be welcome, but IMHO at a later stage once we refine the current code.

Also, I am not sure that `Derive` can play the role of `a constraint on design to help do The Right Thing". I am full for extensions and cool advanced features, but IMHO we may want a bit more validation before conditioning core parts such as STM scheduling.

> I have no idea what the problem is

The first problem is that the feature removed here seems to have been introduced without coordination with the STM designer and design, and worse, without any user at the introduction moment.

As such, it interacts poorly with the rest of the STM and requires unrelated parts to determine "opacity", which they often get wrong. It also introduced code (`make_polimoporhic`) that seems quite strange.

> (and I'm guessing I'm not alone)

That is the reason this and #682 were introduced as RFC. So far, the more I look to the code the more I am convinced we want to move forward with the "proof/stm" unification even at some [temporal] cost on more marginal features, such as `start_dependent_proof` or `Qed exporting`.

> Anything that takes features off of Coq should be very careful in explaining the trade-offs.

To me the tradeoff is hard to compute as so far nobody was taking advantage of those two features TTBOMK.

**Comment by ejgallego** [comment]
@aspiwack some more context, as we have discussed this in the Coq WG but you may have missed it. Some months ago plans started for a "medium" revision to the STM and its API [which has gone basically unchanged since its introduction] in order to provide a better experience for IDE developers and better debugging of paralell processing, etc... and also finish some parts of the integration between the new "execution" model and some other parts of the system.

So far, we have completed the more straighfoward cleanup bits: the toplevel is now a regular STM client, every part of the system uses a single printing part; overlapping parts are less, and some decoupling happened.

After that, it is time to aim for deeper changes, after several meetings among developers and we agreed to try to propose a roadmap in form of a CEP which I am in charge to write, and it is almost done but it did indeed forget to finish it, as I am more or a PRy guy.

Anyways the CEP contained the proposal of single main change: decouple document building from scheduling. Currently, the STM represents documents in a way that reflects both their structure and the intended scheduling [forks, side effect replay, etc...] This is a bit incovenient thou because that representation is leaky for IDEs and either requires complex synchronous protocols, and provides a non-very flexible scheduling in a few cases.

In my view the tradeoff is clear in favor of trying to implement this change, but of course the incremental path is not easy! First, some remaining bits such as #459 should be solved, second, we want indeed to gradually reduce the complexity in the current document model and traversal until we get something similar to the document model UI could profit from. Also, we'd like to express the state at the STM in a functional way; once we reach that point we believe that safe refactoring will be much easier.

So that is the story, regarding the `Opacity` flag, I think it is a good idea, but I think it should be part of a more general mechanism that marks constructions unsuitable to be processed in parallel, so it is restricted by its current scope. But again I insist on the drafty status of these attempts.

**Comment by aspiwack** [comment]
Thanks for the context. It still feels a bit thin to me though. It may just be that I'm a bit out of touch of the current Coq developments (I almost certainly am), but I just don't understand the motivations behind this particular change (I do understand the streamlining of the STM mechanism).

> This doesn't make Derive impossible, it just removes an ad hoc mechanism introduced for it. Derive should just be marked as "never process in parallel", and that'd be good.

Fair. But it was the intention of this flag. Why is this flag not the right way to mark `Derive` as such? What would be better? Why are you removing the flag without implementing the alternative?

> Also, I am not sure that Derive can play the role of `a constraint on design to help do The Right Thing". I am full for extensions and cool advanced features, but IMHO we may want a bit more validation before conditioning core parts such as STM scheduling.

From my point of view, `Derive` is fairly basic. STM is a cool advanced feature. Also, I don't see STM as a "core feature" (I mean, it's certainly a _big_ feature but at the end of the day it's a superficial feature: it is about IDEs. I consider language considerations much more core). But I guess that, in a sense, this is all a matter of point of view.

My point, though, is that however you consider the relative importance of each, looking towards the future and potential extensions is the way to write code in a robust way, and avoid payining for overfitting later.

> The first problem is that the feature removed here seems to have been introduced without coordination with the STM designer and design, and worse, without any user at the introduction moment.

As it happens, this is not the case. I was not capable, at the time, of modifying the STM without Enrico's advice. And there was one user :-p . Of course, in the current workflow I would have had to submit it as a PR, which would have been a better path, but it was, unfortunately, not the way we did things at the time.

> As such, it interacts poorly with the rest of the STM and requires unrelated parts to determine "opacity", which they often get wrong.

You'll have to elaborate (_e.g._ are there examples of code that claimed that they guaranteed opacity and didn't? are there examples of code that claimed they didn't guarantee opacity, but in fact did?). Also, again, what would be a better design and why?

> we want to move forward with the "proof/stm" unification even at some [temporal] cost on more marginal features, such as start_dependent_proof or Qed exporting.

Sure, it makes sense to unify proof and stm (it was always the intention). But I don't understand why it interacts with these features at all.

> But again I insist on the drafty status of these attempts.

Indeed, but I'll argue that a draft without a good explanation is worse than nothing at all. If what you are looking for is comments, you will need to give the readers the tool to comment. From where I stand there is just a feature that I like (I'm biased of course) which seems to get the axe. All I can comment is: « I like the feature don't remove it ». Maybe it would be for the best, but I have no way to assess that.

**Comment by ejgallego** [comment]
> Why is this flag not the right way to mark Derive as such?

Maybe it is just that the name is wrong, but having a "non-opaque" Qed is morally in conflict with Qed, at least in the medium term view. The better solution would be to separate document nodes `VtProofStart` from their intended schedule [this flag].

The key question is "how do we check that a proof can be checked in parallel".

Prior to this patch, the STM used the criteria that "things between a proof starting command and Qed could be checked in parallel" [with some other restrictions that hopefully will go away too"

After this patch, there is an additional criteria that allows things between a proof starting command and Qed not to be checked in parallel. This has a pretty significant impact IMO.

Anyways, the real deal here is "what is a proof starting command" ? We need to solve this question now, currently, responsibilities are shared between different layers depending what the command in question is, with components fighting at some point to take over the responsability.

This is not good and has produced many bugs and hacks to workaround the situation.

> I don't see STM as a "core feature"

Well, these days the STM is in charge of scheduling proof checking, so I'd say that this seems like a pretty core feature to me. It is true however that with recent [some of them pending] changes you can indeed write a very simple toplevel that hooks with vernac and checks a document linearly.

STM it is not just about IDEs, but about building and checking a document. I think that relative importance of stm vs derive is fairly easy to assess, isn't it?

> You'll have to elaborate (e.g. are there examples of code that claimed that they guaranteed opacity and didn't? are there examples of code that claimed they didn't guarantee opacity, but in fact did?). Also, again, what would be a better design and why?

This should be in the CEP, my view is that the less static analysis we have to maintain the better; currently, there is fairly complex static analysis that has to be kept in sync with other parts, and as it turns out, it is not hard to find problems, just see the log in `stm` for some example. Who validates that the static analysis is correct? This bit of information is yet another piece of static analysis that plugin writers have to care about.

Ideally, you could fully get rid of static analysis would we be able to have preemptive concurrency in OCaml, however as this is not the case we will need a bit.

The proposed alternative [still in progress] is to have the scheduler to resort to synchronous execution by default, then recognize that certain patterns are safe to run in parallel, basically effect-free blocks of `Theorem ... Qed`. IMO allowing the extension of this mechanism is not something we want to happen yet.

> why it interacts with these features at all.

Each feature has a point. For `start_dependent_proof`, the reason is to remove some control inversion present currently in the code, and have stricter typing policies on how constants are declared. For `Qed exporting` the reason is to force true opacity again.

For this particular PR, the reason is to enforce the strong invariant that blocks `Proof. ... Qed` are truly opaque and will produce one [possibily mutual] constant.

We'd also like to remove cooking and generation of proof-time side effects [replacing them by a diff + local checking], but that's gonna take more time.

> All I can comment is: « I like the feature don't remove it »

My time for questions has cometh:
 
Why do you like it? What other parts of the system can benefit from it? What are the semantics of a non-opaque `Proof. Qed.` block ? What is the code in `make_polymorphic doing?

Anyways as you have surely seen I've marked this for "more thinking needed", I don't think it is prioritary change as we have like dozens of changes to perform before, but IMHO we should warm the discussion up about `start/close` proof which will be the real deal.

**Comment by gares** [comment]
I see no tension between this PR and the derive plugin.

What makes the derive plugin hard to handle (from the STM pov) is that it overloads the meaning of `Qed`. If it was expecting a different terminator, like `End Derivation`, then things would have been simpler.

I think the root of all evil is the fact that proof terminators are too expressive. A function of type `proof -> unit` can do anything, and among these things break the "opacity" guarantee of the `Qed` keyword.  If proof terminators were simpler, then derive would not have been able to use them and would have simply defined its own terminator, and declare to the STM to process it as it processes `Defined`.

**Comment by ejgallego** [comment]
> I think the root of all evil is the fact that proof terminators are too expressive. A function of type proof -> unit can do anything, and among these things break the "opacity" guarantee of the Qed keyword. If proof terminators were simpler, then derive would not have been able to use them and would have simply defined its own terminator, and declare to the STM to process it as it processes Defined.

+1 indeed we should think what to do in that area.

**Comment by ejgallego** [comment]
Much interesting discussion here, however I don't think this is on the short-term roadmap as the current todo list is largely orthogonal to this change.
